Fix memory leak

jderusse opened the pull request.

Nyholm commented:
Are you sure this will fix it? Php is copy-on-write, which means that it only duplicates a variable in memory if it really has to. It seams like this could be optimized away by the runtime. 

Also. Even if the body is 5 mb large (super large), storing 10mb is not much. 

I still suspect that doing preg_replace on a large body would lead to memory consumption of over 10mb. I can do some testing later if you want.

jderusse commented:
> Php is copy-on-write, 
Exactly, that why, untill we change the value of the variable `$responseContent` the situation is fine. As soon as the variable is altered (by preg_replace or substr), php copy the content of the variable in a new location.

```
$response->setContent('foo'); // $response->content point to the memory 'foo'
$responseContent = $response->getContent(); // both $responseContent AND $response->content point to the memory 'foo'

$responseContent = \preg_replace(...); // $responseContent point to 'foo-replaced', and $response->content point to the memory 'foo' <= twice memory usage

$response->setContent($responseContent); // both $responseContent AND $response->content point to the memory 'foo-replaced'
```


here is a benchmark in the ResponseListener code
```
                if ('text/html' === \substr($response->headers->get('Content-Type'), 0, 9)) {
                    $mp=[];$mu=[];
                    echo '1 ', $mp[] = memory_get_peak_usage(), PHP_EOL;
                    echo '2 ', $mu[] = memory_get_usage(), PHP_EOL;
                    $responseContent = $response->getContent();
                    echo '3 ', $mu[] = memory_get_usage(), ' ('.($mu[count($mu)-1]-$mu[count($mu)-2]).')', PHP_EOL;

                    if (null === $this->newRelicTwigExtension || false === $this->newRelicTwigExtension->isHeaderCalled()) {
                        $responseContent = \preg_replace('/<\s*head\s*>/', '$0'.$this->interactor->getBrowserTimingHeader(), $responseContent);
                    }

                    if (null === $this->newRelicTwigExtension || false === $this->newRelicTwigExtension->isFooterCalled()) {
                        $responseContent = \preg_replace('/<\s*\/\s*body\s*>/', $this->interactor->getBrowserTimingFooter().'$0', $responseContent);
                    }

                    echo '4 ', $mu[] = memory_get_usage(), ' ('.($mu[count($mu)-1]-$mu[count($mu)-2]).')', PHP_EOL;
                    if ($responseContent) {
                        $response->setContent($responseContent);
                    }
                    echo '5 ', $mu[] = memory_get_usage(), ' ('.($mu[count($mu)-1]-$mu[count($mu)-2]).')', PHP_EOL;
                    echo '6 ', $mp[] = memory_get_peak_usage(), ' ('.($mp[count($mp)-1]-$mp[count($mp)-2]).')', PHP_EOL;
                    die;
                }
```
here are the result
```
 1 15311744
 2 15151624
 3 15152000 (376)
 4 15479680 (327680)
 5 15315840 (-163840)
 6 15643520 (331776)
```
as you can see memory increase by +327K (step `4`) after changing the value of `$responseContent` then decrease by 163K when calling `response->setContent` (step `5`)

Now, I changed the code to 
```
                if ('text/html' === \substr($response->headers->get('Content-Type'), 0, 9)) {
                    $mp=[];$mu=[];
                    echo '1 ', $mp[] = memory_get_peak_usage(), PHP_EOL;
                    echo '2 ', $mu[] = memory_get_usage(), PHP_EOL;
                    $responseContent = $response->getContent();
                    echo '3 ', $mu[] = memory_get_usage(), ' ('.($mu[count($mu)-1]-$mu[count($mu)-2]).')', PHP_EOL;
                    $response->setContent('');
                    echo '0 ', $mu[] = memory_get_usage(), ' ('.($mu[count($mu)-1]-$mu[count($mu)-2]).')', PHP_EOL;

                    if (null === $this->newRelicTwigExtension || false === $this->newRelicTwigExtension->isHeaderCalled()) {
                        $responseContent = \preg_replace('/<\s*head\s*>/', '$0'.$this->interactor->getBrowserTimingHeader(), $responseContent);
                    }

                    if (null === $this->newRelicTwigExtension || false === $this->newRelicTwigExtension->isFooterCalled()) {
                        $responseContent = \preg_replace('/<\s*\/\s*body\s*>/', $this->interactor->getBrowserTimingFooter().'$0', $responseContent);
                    }

                    echo '4 ', $mu[] = memory_get_usage(), ' ('.($mu[count($mu)-1]-$mu[count($mu)-2]).')', PHP_EOL;
                    if ($responseContent) {
                        $response->setContent($responseContent);
                    }
                    echo '5 ', $mu[] = memory_get_usage(), ' ('.($mu[count($mu)-1]-$mu[count($mu)-2]).')', PHP_EOL;
                    echo '6 ', $mp[] = memory_get_peak_usage(), ' ('.($mp[count($mp)-1]-$mp[count($mp)-2]).')', PHP_EOL;
                    die;
                }
```
I just added the step `0: $response->setContent('');` after  the getContent

the result are
```
1 15286192
2 15126072
3 15126448 (376)
8 15126448 (0)
4 15290288 (163840)
5 15290288 (0)
6 15454128 (167936)
```
now the step `4` consume only 163K the memory is not uwed twice during the replace.

Notice that the memory peak increased by 167K, wheras the previous implementation was 331K

Nyholm commented:
You are correct. This will fix the issue. 

For the record, I did a small test. My method did not help. 

```php
<?php
$body = file_get_contents('https://raw.githubusercontent.com/twbs/bootstrap/v4-dev/dist/css/bootstrap.css');

$mp=[];$mu=[];
echo '1 ', $mp[] = memory_get_peak_usage(), PHP_EOL;
echo '2 ', $mu[] = memory_get_usage(), PHP_EOL;

//preg_replace('|2px r|','xxx', $body);
$pos = strripos($body, '2px r');
$replacement = 'xxx';
$content = substr($body, 0, $pos).$replacement.substr($body, $pos);

echo '3 ', $mu[] = memory_get_usage(), ' ('.($mu[count($mu)-1]-$mu[count($mu)-2]).')', PHP_EOL;
echo '4 ', $mp[] = memory_get_peak_usage(), ' ('.($mp[count($mp)-1]-$mp[count($mp)-2]).')',  PHP_EOL;
```

```
▶ php test.php (with preg_replace)
1 589920
2 573424
3 573800 (376)
4 749944 (160024)


▶ php test.php (with substr)
1 591096
2 574600
3 751104 (176504)
4 853504 (262408)
```

jderusse commented:
You have a memory increase because you use distinguished variables
```
$pos = strripos($body, '2px r');
$replacement = 'xxx';

// 2 memory slots: $content and $body 
$content = substr($body, 0, $pos).$replacement.substr($body, $pos);

// 1 memory slot: old value of body is released.
$body = substr($body, 0, $pos).$replacement.substr($body, $pos);
```
But it's weird that the memory_peak is increased. I think that before concatenating the final string, every part of the string are copyied in memory, whereas preg_replace replace the variable itself...

I've method few change to still keep preg_replace, but avoid the `return null` when the pattern doesnot match

Nyholm commented on Listener/ResponseListener.php:
```diff
@@ -85,18 +85,17 @@ public function onKernelResponse(FilterResponseEvent $event): void
 
                     if (null === $this->newRelicTwigExtension || false === $this->newRelicTwigExtension->isHeaderCalled()) {
-                        $responseContent = \preg_replace('/<\s*head\s*>/', '$0'.$this->interactor->getBrowserTimingHeader(), $responseContent);
+                        $responseContent = \preg_replace('|<head>|i', '$0'.$this->interactor->getBrowserTimingHeader(), $responseContent);
```
Im fine with changing the regex here (since we do 2.0). If there is causing any issue for someone then they will provide a better test.
